tools/multithreaded-writer/TimestreamMultithreadedWriter/src/main/java/com/amazonaws/sample/timestream/multithreaded/TimestreamWriterWorker.java

```diff
@@ -0,0 +1,169 @@
+    private final TimestreamWriteClient writeClient;
+    private final AtomicInteger writesInFlight;
+    private final TimestreamInsertionMetrics insertionMetrics;
+    private final TimestreamInitializer timestreamInitializer;
+    private final long maxRetryDurationMs;
+
+    public TimestreamWriterWorker(@NonNull final BlockingQueue<WriteRecordsRequest> queue,
+                                  @NonNull final TimestreamWriterConfig writerConfig,
+                                  @NonNull final AtomicInteger writesInFlight,
+                                  @NonNull final TimestreamInsertionMetrics insertionMetrics,
+                                  @NonNull final TimestreamInitializer timestreamInitializer) {
+        this.queue = queue;
+        this.writeClient = writerConfig.getWriteClient();
+        this.maxRetryDurationMs = writerConfig.getMaxRetryDurationMs();
+        this.writesInFlight = writesInFlight;
+        this.insertionMetrics = insertionMetrics;
+        this.timestreamInitializer = timestreamInitializer;
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (isRunning.get()) {
+                final WriteRecordsRequest writeRecord = queue.poll(POLL_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+                if (writeRecord == null) // Empty poll? Try again.
+                    continue;
+                // If isWriteApproximatelyComplete checks happens here, it will return true,
+                // where it should return false. This is ok if we need approximate check.
+                writesInFlight.incrementAndGet();
+                try {
+                    safelyProcessWriteRecord(writeRecord);
+                } finally {
+                    writesInFlight.decrementAndGet();
+                }
+            }
+        } catch (final InterruptedException e) {
+            LOG.info("Current thread was interrupted. Exiting.");
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private void safelyProcessWriteRecord(@NonNull final WriteRecordsRequest writeRecord) {
+        try {
+            final long recordPollTime = System.currentTimeMillis();
+            do {
+                final TimestreamInsertionMetrics currentWriteMetrics = insertToTimestream(writeRecord);
+                // if we won't retry inserting this record, capture write latency
+                if (currentWriteMetrics.getNonSDKReties().get() == 0) {
+                    currentWriteMetrics.recordLatencyMs(System.currentTimeMillis() - recordPollTime);
+                    insertionMetrics.accumulate(currentWriteMetrics);
+                    break;
+                }
+
+                // In case of retryable errors we want to retry inserting the record to Timestream,
+                // in addition to SDK retry up to MAX_RETRY_DURATION_MS.
+                final long recordMillisAge = System.currentTimeMillis() - recordPollTime;
+                if (recordMillisAge > maxRetryDurationMs) {
```

Comment:
default maxRetryDuration is set to 5mins. Is that too high? Does that mean this will continuously retry a single request for up to 5mins? Also should we have some backoff mechanism for the retry?

Reply:
Regarding backoff mechanism: We don't need it - I'm using the default recommended SDK write client and it already has backoff mechanism - see `TimestreamWriterConfig.defaultRecommendedWriteClient`.

Regarding `maxRetryDuration` value - this depends on the use-case and writer/customer requirements (for example - what should be the system behavior in case of an outage?). This is by purpose defined as an variable, so it can be easily adjusted through `TimestreamWriterConfig`, and is even exposed in the lambda sample as an environment variable: `MAX_TIMESTREAM_RECORD_INSERTION_RETRY_DURATION_S`.